nav2_costmap_2d/src/clear_costmap_service.cpp

```diff
@@ -0,0 +1,142 @@
+      std::bind(&ClearCostmapService::clearExceptRegionCallback, this,
+      std::placeholders::_1, std::placeholders::_2, std::placeholders::_3));
+}
+
+void ClearCostmapService::clearExceptRegionCallback(
+  const shared_ptr<rmw_request_id_t>/*request_header*/,
+  const shared_ptr<ClearExceptRegion::Request> request,
+  const shared_ptr<ClearExceptRegion::Response>/*response*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "Received request to clear " + costmap_->getName());
+
+  if (costmap_ == nullptr) {
+    RCLCPP_ERROR(node_->get_logger(), "Costmap is undefined. Doing nothing.");
+    return;
+  }
+
+  RCLCPP_INFO(node_->get_logger(), "Will proceed with clearing the costmap");
+
+  clearExceptRegion(request->reset_distance);
+}
+
+void ClearCostmapService::clearExceptRegion(const double reset_distance)
+{
+  double x, y;
+
+  if (!getPose(x, y)) {
+    RCLCPP_ERROR(node_->get_logger(), "Cannot clear map because robot pose cannot be retrieved.");
+    return;
+  }
+
+  auto layers = costmap_->getLayeredCostmap()->getPlugins();
+
+  for (auto & layer : *layers) {
+    if (isClearable(getLayerName(*layer))) {
+      auto costmap_layer = std::static_pointer_cast<CostmapLayer>(layer);
+      clearLayerExceptRegion(costmap_layer, x, y, reset_distance);
+    }
+  }
+}
+
+bool ClearCostmapService::isClearable(const string & layer_name) const
+{
+  return any_of(begin(clearable_layers_), end(clearable_layers_),
+           [&layer_name](auto l) {return l == layer_name;});
+}
+
+void ClearCostmapService::clearLayerExceptRegion(
+  shared_ptr<CostmapLayer> & costmap, double pose_x, double pose_y, double reset_distance)
+{
+  std::unique_lock<Costmap2D::mutex_t> lock(*(costmap->getMutex()));
+
+  double start_point_x = pose_x - reset_distance / 2;
+  double start_point_y = pose_y - reset_distance / 2;
+  double end_point_x = start_point_x + reset_distance;
+  double end_point_y = start_point_y + reset_distance;
+
+  int start_x, start_y, end_x, end_y;
+  costmap->worldToMapNoBounds(start_point_x, start_point_y, start_x, start_y);
+  costmap->worldToMapNoBounds(end_point_x, end_point_y, end_x, end_y);
+
+  unsigned int size_x = costmap->getSizeInCellsX();
+  unsigned int size_y = costmap->getSizeInCellsY();
+
+  // Clearing the four rectangular regions around the one we want to keep
+  // top region
+  costmap->resetMapToValue(0, 0, size_x, start_y, NO_INFORMATION);
+  // left region
+  costmap->resetMapToValue(0, start_y, start_x, end_y, NO_INFORMATION);
+  // right region
+  costmap->resetMapToValue(end_x, start_y, size_x, end_y, NO_INFORMATION);
+  // bottom region
+  costmap->resetMapToValue(0, end_y, size_x, size_y, NO_INFORMATION);
+
+  double ox = costmap->getOriginX(), oy = costmap->getOriginY();
+  double width = costmap->getSizeInMetersX(), height = costmap->getSizeInMetersY();
+  costmap->addExtraBounds(ox, oy, ox + width, oy + height);
+}
+
+bool ClearCostmapService::getPose(double & x, double & y) const
```

Comment:
Suggestion:  Pose contains both position and orientation. I suggest to rename the method to getPosition since it only updates 2d position.

Reply:
Good idea, I'll rename it.